search/views.py

```diff
@@ -0,0 +1,127 @@
+        if not term:
+            raise Http404('Define "term" get parameter')
+
+        # if we have entity with id == int(term)
+        # then redirect to this entity
+        search_entity = self.redirect_search_entity.search_by_redirect_field(term)
+        if search_entity:
+            return redirect(search_entity.url, permanent=True)
+
+        limiter = search_engine.SearchLimiter(self.limit, self.search_entities)
+        limited_data = limiter.search_and_limit_data(term)
+
+        # if there is only one autocompleted entity
+        # then redirect to this entity
+        if limiter.size == 1:
+            return redirect(list(limited_data.values())[0][0].url, permanent=True)
+
+        self.object = self.get_object()
+        template = self.template_path.format(
+            'results' if limited_data else 'no_results')
+
+        context = self.get_context_data(object=self.object)
+        context.update({
+            **limited_data,
+            'term': term
+        })
+
+        return render(request, template, context)
+
+
+class Autocomplete(View):
+
+    limit = 20
+    search_entities = []  # query lookups
+    entity_fields = {}  # model fields that will be contained in the autocomplete item
+
+    search_url = reverse_lazy(Page.CUSTOM_PAGES_URL_NAME, kwargs={'page': 'search'})
+    see_all_label = 'See all results'
+
+    def prepare_fields(self, items, entity, context) -> list:
+        def get_dict(item):
+            fields = {attr: getattr(item, attr) for attr in entity}
+            fields.update(context)
+            return fields
+
+        return [get_dict(item) for item in items]
+
+    def see_all_link_as_list(self, term):
+        url = '{}?term={}'.format(self.search_url, term)
+        return [{'name': self.see_all_label,
+                 'url': url,
+                 'type': 'see_all'}]
+
+    def get(self, request):
+        term = request.GET.get('term')
+
+        limiter = search_engine.SearchLimiter(self.limit, self.search_entities)
+        limited_data = limiter.search_and_limit_data(term)
+
+        if not limited_data:
+            return JsonResponse({})
+
+        autocomplete_items = (
+            self.prepare_fields(
+                limited_data.get('category', []),
+                self.entity_fields['category'],
+                {'type': 'category'}
+            ) +
+            self.prepare_fields(
+                limited_data.get('product', []),
+                self.entity_fields['product'],
+                {'type': 'product'}
+            ) +
+            self.see_all_link_as_list(request.GET.get('term'))
+        )
+
+        return JsonResponse(autocomplete_items, safe=False)
+
+
+class AdminAutocomplete(View):
+
+    limit = 20
+    search_entities = []
+
+    def get(self, request):
+        term, page_type = request.GET.get('term'), request.GET.get('pageType')
+        if page_type not in [entity.name for entity in self.search_entities]:
```

Comment:
current_search = next((item for item in self.search_entities if item.name == page_type), None)
if current_search is None:
    return